### PR TITLE
[NNAPI EP] Enable Slice support

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: [NNAPI EP] Enable Slice support

**Motivation and Context**
- Enabled slice support which is part of SSD models
- Moved the shape inference functions from CPU slice impl to 2 separated helper files, which can be shared without including certain dependencies of CPU EP such as eigen
- Added ends_ field for SliceOp::PrepareForComputeMetadata
- Updated UT

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/slice.cc [hunk 1]
```diff
@@ -12,6 +12,8 @@
 #include "core/providers/op_kernel_type_control.h"
 #include "core/providers/op_kernel_type_control_utils.h"
 
+#include "slice_helper.h"
```
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/tensor/slice.cc:15]: general: can we use the full path of the included file?

File: onnxruntime/core/providers/cpu/tensor/slice_compute_metadata.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// This file contains the defnition of the PrepareForComputeMetadata for Slice operator
```
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/tensor/slice_compute_metadata.h:4]: >defnition [](http://example.com/codeflow?start=26&length=9)

"defnition" -> "definition"

File: onnxruntime/core/providers/cpu/tensor/slice_compute_metadata.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#pragma once
```
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/tensor/slice_compute_metadata.h:5]: can the headers be made self-contained? e.g., include <vector> in this file
[https://google.github.io/styleguide/cppguide.html#Self_contained_Headers](https://google.github.io/styleguide/cppguide.html#Self_contained_Headers)

File: onnxruntime/core/providers/cpu/tensor/slice_compute_metadata.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace onnxruntime {
+
+namespace SliceOp {
+struct PrepareForComputeMetadata {
+  PrepareForComputeMetadata() = delete;
+  PrepareForComputeMetadata(const std::vector<int64_t>& input_dimensions)
```
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/tensor/slice_compute_metadata.h:12]: can this be `explicit`?

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc [hunk 5]
```diff
@@ -2561,6 +2561,179 @@ Status EluOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const No
...
+  };
+
+  ORT_RETURN_IF_ERROR(AddOperand("starts", param_dimen, compute_metadata.starts_));  //nnapi_begin
+
+  // NNAPI has 2 slice operations
+  // - ANEURALNETWORKS_SLICE
+  //    Simpler and faster version of slice without steps, available from ANEURALNETWORKS_FEATURE_LEVEL_3
+  //    Use this one if no step other than 1 is used in ONNX slice
+  // - ANEURALNETWORKS_STRIDED_SLICE
+  //    More comprehensice version, available from ANEURALNETWORKS_FEATURE_LEVEL_2
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:2694]: >comprehensice [](http://example.com/codeflow?start=13&length=13)

"comprehensice" -> "comprehensive"

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc [hunk 6]
```diff
@@ -2644,6 +2817,7 @@ static OpBuilderRegistrations CreateOpBuilderRegistrations() {
   }
 
   NNAPI_EP_ADD_SINGLE_OP_BUILDER("Elu", EluOpBuilder);
+  NNAPI_EP_ADD_SINGLE_OP_BUILDER("Slice", SliceOpBuilder);
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:2820]: >NNAPI_EP_ADD_SINGLE_OP_BUILDER [](http://example.com/codeflow?start=2&length=30)

why are some of these `NNAPI_EP_ADD_SINGLE_OP_BUILDER`'s in separate blocks?

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc [hunk 7]
```diff
@@ -1639,6 +1654,62 @@ class EluOpSupportChecker : public BaseOpSupportChecker {
 
 #pragma endregion
 
+#pragma region op_slice
+
+class SliceOpSupportChecker : public BaseOpSupportChecker {
+ private:
+  int32_t GetMinSupportedNNAPIFeatureLevel(const Node& /* node */, const OpSupportCheckParams& /* params */) const override {
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc:1661]: nit: there were a bunch of other changes to break similar lines, but this line is a bit long

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc [hunk 8]
```diff
@@ -1639,6 +1654,62 @@ class EluOpSupportChecker : public BaseOpSupportChecker {
...
+
+  if (input_shape.size() > 4) {
+    LOGS_DEFAULT(VERBOSE) << "Slice only supports 1-4d shape, input is "
+                          << input_shape.size() << "d shape";
+    return false;
+  }
+
+  // TODO, replace with std::find when we switch to c++17
+  if (std::any_of(input_shape.cbegin(), input_shape.cend(), [](int32_t i) { return i == 0; })) {
+    LOGS_DEFAULT(VERBOSE) << "Slice doesn't dynamic input shape";
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc:1686]: "doesn't support"

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc (hunk 6 above)
Comment (gwang-msft) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:2820]: There are just in order of when the op support was added.
Changed to sorted order

File: onnxruntime/core/providers/cpu/tensor/slice.h [hunk 9]
```diff
@@ -1,33 +1,16 @@
...
-struct PrepareForComputeMetadata {
-  PrepareForComputeMetadata() = delete;
-  PrepareForComputeMetadata(const std::vector<int64_t>& input_dimensions)
-      : input_dimensions_(input_dimensions) {
-    size_t dimension_count = input_dimensions.size();
-    starts_.resize(dimension_count, 0);
-    steps_.resize(dimension_count, 1);
-    output_dims_ = input_dimensions;
-  }
+#include "slice_compute_metadata.h"
```
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/tensor/slice.h:11]: >"slice_compute_metadata.h" [](http://example.com/codeflow?start=9&length=26)

full path?

File: onnxruntime/core/providers/cpu/tensor/slice_helper.h [hunk 10]
```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// This file contains the functions compute the starts, steps (strides) and output shape
+// for Slice op, which can be called from other ops or EPs.
+#pragma once
+#include "slice_compute_metadata.h"
```
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/tensor/slice_helper.h:7]: >"slice_compute_metadata.h" [](http://example.com/codeflow?start=9&length=26)

full path?

**Comment by edgchen1** [comment]
> const T& clamp(const T& v, const T& lo, const T& hi) {

is this still needed in this file?

---
Refers to: onnxruntime/core/providers/cpu/tensor/slice.cc:51 in a6155a4. [](commit_id = a6155a4301ad3ce4a863e98bf4c60d17252ae189, deletion_comment = False)

File: onnxruntime/core/providers/cpu/tensor/slice_helper.h [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+      compute_metadata.output_dims_[axis] = 0;
+    else
+      compute_metadata.output_dims_[axis] = temp;
+  }
+
+  return Status::OK();
+}
+
+// compute output_dims with steps (Slice V10)
+// Please not this will not Flatten the output shape
```
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/tensor/slice_helper.h:69]: >not [](http://example.com/codeflow?start=10&length=3)

"Please note"

Comment (edgchen1) [reply onnxruntime/core/providers/cpu/tensor/slice_helper.h:69]: also shows up in other PrepareForCompute()

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc (hunk 6 above)
Comment (edgchen1) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc]: I was wondering why we have
```
NNAPI_EP_ADD_SHARED_OP_BUILDER(...);
{
  NNAPI_EP_ADD_SHARED_OP_BUILDER(...);
}
NNAPI_EP_ADD_SHARED_OP_BUILDER(...);
```

**Comment by gwang-msft** [comment]
> > const T& clamp(const T& v, const T& lo, const T& hi) {
> 
> is this still needed in this file?
> 
> Refers to: onnxruntime/core/providers/cpu/tensor/slice.cc:51 in [a6155a4](https://github.com/microsoft/onnxruntime/commit/a6155a4301ad3ce4a863e98bf4c60d17252ae189). [](commit_id = [a6155a4](https://github.com/microsoft/onnxruntime/commit/a6155a4301ad3ce4a863e98bf4c60d17252ae189), deletion_comment = False)

Removed

**Review by edgchen1 — APPROVED** [review]
:shipit:

**Pull request merged**